services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy

```diff
@@ -0,0 +1,365 @@
+
+        when: "Create a QinQ flow"
+        def flow = flowHelperV2.randomFlow(swP)
+        flow.source.vlanId = srcVlanId
+        flow.source.innerVlanId = srcInnerVlanId
+        flow.destination.vlanId = dstVlanId
+        flow.destination.innerVlanId = dstInnerVlanId
+        def response = flowHelperV2.addFlow(flow)
+
+        then: "Response contains correct info about innerVlanIds"
+        with(response) {
+            it.source.vlanId == (srcVlanId ? srcVlanId : srcInnerVlanId)
+            it.source.innerVlanId == (srcVlanId ? srcInnerVlanId : 0)
+            it.destination.vlanId == (dstVlanId ? dstVlanId : dstInnerVlanId)
+            it.destination.innerVlanId == (dstVlanId ? dstInnerVlanId : 0)
+        }
+
+        and: "Flow is really created with requested innerVlanIds"
+        with(northbound.getFlow(flow.flowId)) {
+            it.source.vlanId == (srcVlanId ? srcVlanId : srcInnerVlanId)
+            it.source.innerVlanId == (srcVlanId ? srcInnerVlanId : 0)
+            it.destination.vlanId == (dstVlanId ? dstVlanId : dstInnerVlanId)
+            it.destination.innerVlanId == (dstVlanId ? dstInnerVlanId : 0)
+        }
+
+        // is not implemented yet
+//        and: "Flow is valid and pingable"
+//        northbound.validateFlow(flow.flowId).each { assert it.asExpected }
+//        verifyAll(northbound.pingFlow(flow.flowId, new PingInput())) {
+//            it.forward.pingSuccess
+//            it.reverse.pingSuccess
+//        }
+
+        and: "The flow allows traffic"
+        def traffExam = traffExamProvider.get()
+        def examQinQFlow = new FlowTrafficExamBuilder(topology, traffExam)
+                .buildBidirectionalExam(flowHelperV2.toV1(flow), 1000, 5)
+        withPool {
+            [examQinQFlow.forward, examQinQFlow.reverse].eachParallel { direction ->
+                def resources = traffExam.startExam(direction)
+                direction.setResources(resources)
+                assert traffExam.waitExam(direction).hasTraffic()
+            }
+        }
+
+        // is not implemented yet
+//        and: "Involved switches pass switch validation"
+//        def involvedSwitchesFlow1 = pathHelper.getInvolvedSwitches(
+//                pathHelper.convert(northbound.getFlowPath(flow.flowId))
+//        )
+//        involvedSwitchesFlow1.each {
+//            with(northbound.validateSwitch(it.dpId)) { validation ->
+//                validation.verifyRuleSectionsAreEmpty(["missing", "excess", "misconfigured"])
+//                validation.verifyMeterSectionsAreEmpty(["missing", "excess", "misconfigured"])
+//            }
+//        }
+
+        when: "Create a simple flow on the same port"
+        def flow2 = flowHelper.randomFlow(swP).tap {
+            it.source.portNumber = flow.source.portNumber
+            it.source.vlanId = flow.source.vlanId - 1
+            it.destination.portNumber = flow.destination.portNumber
+            it.destination.vlanId = flow.destination.vlanId - 1
```

Comment:
How about trying to use inner vlan of the qnq flow here or sth more tricky? I don't really see what can go wrong here

Reply:
according to the comment above
it is going to conflict with QinQ flow